src/Ema/Example/Ex04_Multi.hs

```diff
@@ -42,38 +52,60 @@ main = do
 -- TODO: Can we do generic of this too?
 -- Can demo in 'mergeSite' (of two Emanotes?)
 instance EmaSite R where
-  siteInput cliAct enc () = do
-    x1 :: Dynamic m Ex02.Model <- siteInput cliAct (innerRouteEncoder (_As @"R_Clock") enc) Ex02.delayNormal
-    x2 :: Dynamic m Ex03.Model <- siteInput cliAct (innerRouteEncoder (_As @"R_Store") enc) ()
-    pure $ liftA2 (\x y -> I x :* I y :* Nil) x1 x2
-  siteOutput enc m = \case
-    R_Index ->
-      Ema.AssetGenerated Ema.Html $ renderIndex enc m
-    -- TODO: Can all of these be generalized? (constructor with 1 encoder; delegate)
-    R_Hello r ->
-      siteOutput (innerRouteEncoder (_As @"R_Hello") enc) (innerModel m) r
-    R_Basic r ->
-      siteOutput (innerRouteEncoder (_As @"R_Basic") enc) (innerModel m) r
-    R_Store r ->
-      siteOutput (innerRouteEncoder (_As @"R_Store") enc) (innerModel m) r
-    R_Clock r ->
-      siteOutput (innerRouteEncoder (_As @"R_Clock") enc) (innerModel m) r
+  siteInput cliAct () = do
+    x1 :: Dynamic m Ex02.Model <- siteInput @Ex02.Route cliAct Ex02.delayNormal
+    x2 :: Dynamic m Ex02.Model <- siteInput @Ex02.Route cliAct Ex02.delayFast
+    x3 :: Dynamic m Ex03.Model <- siteInput @Ex03.Route cliAct ()
+    pure $ liftA3 M x1 x2 x3
+  siteOutput enc m sr =
+    let I () :* I m2 :* I m3 :* I m4 :* I m5 :* I m6 :* Nil = subModels @R m
+     in case sr of
+          R_Index ->
+            Ema.AssetGenerated Ema.Html $ renderIndex enc m
+          -- TODO: Can all of these be generalized? (constructor with 1 encoder; delegate)
+          R_Hello r ->
+            -- TODO: add `subSiteOutput enc m (_As @"R_Hello") r`?
+            siteOutput (innerEncoderWithModel m (_As @"R_Hello") enc) m2 r
+          R_Basic r ->
+            siteOutput (innerEncoderWithModel m (_As @"R_Basic") enc) m3 r
+          R_Clock r ->
+            siteOutput (innerEncoderWithModel m (_As @"R_Clock") enc) m4 r
+          R_ClockFast r ->
+            siteOutput (innerEncoderWithModel m (_As @"R_Clock") enc) m5 r
+          R_Store r ->
+            siteOutput (innerEncoderWithModel m (_As @"R_Store") enc) m6 r
+
+{- | Return the inner `RouteEncoder` corresponding to the route `i`, assuming
+ that the model value `a` is known ahead of time.
+-}
+innerEncoderWithModel ::
+  forall (pr :: OpticKind) o i a b.
+  pr `Is` A_Prism =>
+  a ->
+  Optic' pr NoIx o i ->
+  RouteEncoder a o ->
+  RouteEncoder b i
+innerEncoderWithModel m innerRoute enc =
+  enc
+    & mapRouteEncoderRoute innerRoute
+    & mapRouteEncoderModel (\_ -> m) -- TODO: this is wrong; why discard inner values?
```

Comment:
This part is icky. Inside of `siteOutput` one normally calls `encodeRoute` with the same model passed to `siteOutput` (in which case this function would behave correctly), but will the user ever call `encodeRoute` with a *different* model value?

Reply:
I think it should not. That's why I wrote it like this in `abacateiro`:
```hs
subRouteEncoder ::
  forall ctor field m n i o.
  HasField' field m n =>
  AsConstructor' ctor i o =>
  m ->
  RouteEncoder m i ->
  RouteEncoder n o
subRouteEncoder m =
  mapRouteEncoder equality (_Ctor' @ctor) (flip (setField @field) m)
```
Do do something like this in this branch, I think `HasSubModels` should be a "lens" instead of just a "getter".

Reply:
If it's changed to a lens, the main problem is then how to identify which of the subroutes we want to "get" and get the corresponding model. Can we access the n-th term of the corresponding `NP I` by using the constructor name? Perhaps such a function can be written by induction if we put it in a typeclass?

With the model being a record I could avoid such problems by using the field names, as noted above.

Reply:
Another much easier option is to limit what the user can do and only pass as argument the Prism' (the route encoder applied to the model) 🙃